### PR TITLE
Fixes to tasking from Tacho optimization days

### PR DESCRIPTION
**Pull request by dhollman**

### CODE REVIEW

**Comment by ibaned** [comment]
Getting the following in an attempted spot-check:
```
/ascldap/users/daibane/src/kokkos/core/src/impl/Kokkos_SimpleTaskScheduler.hpp:103:3: error: ‘constexpr Kokkos::Impl::ExecutionSpaceInstanceStorage<ExecutionSpace>::ExecutionSpaceInstanceStorage()’ cannot be overloaded
   ExecutionSpaceInstanceStorage()
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/ascldap/users/daibane/src/kokkos/core/src/impl/Kokkos_SimpleTaskScheduler.hpp:97:3: error: with ‘constexpr Kokkos::Impl::ExecutionSpaceInstanceStorage<ExecutionSpace>::ExecutionSpaceInstanceStorage()’
   ExecutionSpaceInstanceStorage()
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/ascldap/users/daibane/src/kokkos/core/src/impl/Kokkos_SimpleTaskScheduler.hpp:154:3: error: ‘Kokkos::Impl::MemorySpaceInstanceStorage<MemorySpace>::MemorySpaceInstanceStorage()’ cannot be overloaded
   MemorySpaceInstanceStorage()
   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/ascldap/users/daibane/src/kokkos/core/src/impl/Kokkos_SimpleTaskScheduler.hpp:149:3: error: with ‘Kokkos::Impl::MemorySpaceInstanceStorage<MemorySpace>::MemorySpaceInstanceStorage()’
   MemorySpaceInstanceStorage()
   ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by dhollman** [comment]
Yeah, from some reason the rebase added both.  I fixed it.  No idea how that happened

**Comment by ibaned** [comment]
Haha I was wondering why anyone would write both. Chalk it up to Git's simple merge algorithms. re-running.

File: core/src/Cuda/Kokkos_Cuda_Atomic_Intrinsics.hpp [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+    if (b == __ATOMIC_SEQ_CST) return __ATOMIC_SEQ_CST;
+    if (b == __ATOMIC_RELAXED) return a;
+    switch (a) {
+    case __ATOMIC_SEQ_CST:
+    case __ATOMIC_ACQ_REL: return a;
+    case __ATOMIC_CONSUME:
+    case __ATOMIC_ACQUIRE: if (b != __ATOMIC_ACQUIRE) return __ATOMIC_ACQ_REL; else return __ATOMIC_ACQUIRE;
+    case __ATOMIC_RELEASE: if (b != __ATOMIC_RELEASE) return __ATOMIC_ACQ_REL; else return __ATOMIC_RELEASE;
+    case __ATOMIC_RELAXED: return b;
+    default: assert(0);
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Atomic_Intrinsics.hpp]: Missing `#include <cassert>`

Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Atomic_Intrinsics.hpp]: Shouldn't you be calling `Kokkos::abort()` instead?

Comment (ibaned) [reply core/src/Cuda/Kokkos_Cuda_Atomic_Intrinsics.hpp]: I don't think `Kokkos::abort()` can be called from device code? better to just include `<cassert>`.

**Comment by ibaned** [comment]
Spot check actually passed:
```
Running on machine: kokkos-dev
Repository Status:  7594d0945dc6ba90a1895087ac5fc03cc9f0dfdf fixed a bug from the rebase?


Going to test compilers:  gcc/5.3.0 gcc/7.3.0 intel/17.0.1 clang/4.0.1 cuda/8.0.44
Testing compiler gcc/5.3.0
  Starting job gcc-5.3.0-OpenMP-release
  PASSED gcc-5.3.0-OpenMP-release
Testing compiler gcc/7.3.0
  Starting job gcc-5.3.0-OpenMP-hwloc-release
  PASSED gcc-5.3.0-OpenMP-hwloc-release
  Starting job gcc-7.3.0-Serial-release
  PASSED gcc-7.3.0-Serial-release
Testing compiler intel/17.0.1
  Starting job gcc-7.3.0-Serial-hwloc-release
  PASSED gcc-7.3.0-Serial-hwloc-release
  Starting job intel-17.0.1-OpenMP-release
  PASSED intel-17.0.1-OpenMP-release
Testing compiler clang/4.0.1
  Starting job intel-17.0.1-OpenMP-hwloc-release
  PASSED intel-17.0.1-OpenMP-hwloc-release
  Starting job clang-4.0.1-Pthread_Serial-release
  PASSED clang-4.0.1-Pthread_Serial-release
Testing compiler cuda/8.0.44
  Starting job clang-4.0.1-Pthread_Serial-hwloc-release
  PASSED clang-4.0.1-Pthread_Serial-hwloc-release
  Starting job cuda-8.0.44-Cuda_OpenMP-release
  PASSED cuda-8.0.44-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-4.0.1-Pthread_Serial-hwloc-release build_time=220 run_time=485
clang-4.0.1-Pthread_Serial-release build_time=212 run_time=617
cuda-8.0.44-Cuda_OpenMP-release build_time=596 run_time=540
gcc-5.3.0-OpenMP-hwloc-release build_time=191 run_time=124
gcc-5.3.0-OpenMP-release build_time=192 run_time=127
gcc-7.3.0-Serial-hwloc-release build_time=161 run_time=305
gcc-7.3.0-Serial-release build_time=161 run_time=229
intel-17.0.1-OpenMP-hwloc-release build_time=457 run_time=212
intel-17.0.1-OpenMP-release build_time=465 run_time=226
```

File: core/src/Cuda/Kokkos_Cuda_Atomic_Intrinsics.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Atomic_Intrinsics.hpp]: You can.  See https://github.com/kokkos/kokkos/blob/d3a941925cbfb71785d8ea68259123ed52d3f9da/core/src/impl/Kokkos_Error.hpp#L73-L82 and https://github.com/kokkos/kokkos/blob/d3a941925cbfb71785d8ea68259123ed52d3f9da/core/src/Cuda/Kokkos_Cuda_abort.hpp#L69-L81

**Comment by ibaned** [comment]
final spot check finally done
```
Running on machine: kokkos-dev
Repository Status:  bf8255d3cd5704914ce3aae465d4190c016c88e4 Try adding cassert


Going to test compilers:  gcc/5.3.0 gcc/7.3.0 intel/17.0.1 clang/4.0.1 cuda/8.0.44
Testing compiler gcc/5.3.0
  Starting job gcc-5.3.0-OpenMP-release
  PASSED gcc-5.3.0-OpenMP-release
Testing compiler gcc/7.3.0
  Starting job gcc-5.3.0-OpenMP-hwloc-release
  PASSED gcc-5.3.0-OpenMP-hwloc-release
  Starting job gcc-7.3.0-Serial-release
  PASSED gcc-7.3.0-Serial-release
Testing compiler intel/17.0.1
  Starting job gcc-7.3.0-Serial-hwloc-release
  PASSED gcc-7.3.0-Serial-hwloc-release
  Starting job intel-17.0.1-OpenMP-release
  PASSED intel-17.0.1-OpenMP-release
Testing compiler clang/4.0.1
  Starting job intel-17.0.1-OpenMP-hwloc-release
  PASSED intel-17.0.1-OpenMP-hwloc-release
  Starting job clang-4.0.1-Pthread_Serial-release
  PASSED clang-4.0.1-Pthread_Serial-release
Testing compiler cuda/8.0.44
  Starting job clang-4.0.1-Pthread_Serial-hwloc-release
  PASSED clang-4.0.1-Pthread_Serial-hwloc-release
  Starting job cuda-8.0.44-Cuda_OpenMP-release
  PASSED cuda-8.0.44-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-4.0.1-Pthread_Serial-hwloc-release build_time=239 run_time=412
clang-4.0.1-Pthread_Serial-release build_time=213 run_time=667
cuda-8.0.44-Cuda_OpenMP-release build_time=766 run_time=700
gcc-5.3.0-OpenMP-hwloc-release build_time=188 run_time=116
gcc-5.3.0-OpenMP-release build_time=232 run_time=175
gcc-7.3.0-Serial-hwloc-release build_time=281 run_time=213
gcc-7.3.0-Serial-release build_time=200 run_time=331
intel-17.0.1-OpenMP-hwloc-release build_time=491 run_time=190
intel-17.0.1-OpenMP-release build_time=541 run_time=265
```

**Pull request merged**